lib/Service/AvatarService.php

```diff
@@ -0,0 +1,174 @@
+		$allowedMimeTypes = [
+			'image/jpeg',
+			'image/png',
+		];
+		if (!in_array($mimeType, $allowedMimeTypes)) {
+			throw new InvalidArgumentException($this->l->t('Unknown filetype'));
+		}
+
+		try {
+			$folder = $this->appData->getFolder('room-avatar');
+		} catch (NotFoundException $e) {
+			$folder = $this->appData->newFolder('room-avatar');
+		}
+		$token = $room->getToken();
+		$content = $image->data();
+		$folder->newFile($token, $content);
+		$this->cache->set($token . '.avatarVersion', md5($content));
+	}
+
+	public function getAvatar(Room $room, ?IUser $user): ISimpleFile {
+		$token = $room->getToken();
+		try {
+			$folder = $this->appData->getFolder('room-avatar');
+			if ($folder->fileExists($token)) {
+				$file = $folder->getFile($token);
+			}
+		} catch (NotFoundException $e) {
+		}
+		// Fallback
+		if (!isset($file)) {
+			if ($room->getType() === Room::TYPE_ONE_TO_ONE) {
+				$users = json_decode($room->getName(), true);
+				foreach ($users as $participantId) {
+					if ($participantId !== $user->getUID()) {
+						$avatar = $this->avatarManager->getAvatar($participantId);
+						$file = $avatar->getFile(512);
+					}
+				}
+			} elseif ($room->getObjectType() === 'file') {
+				$file = new InMemoryFile($token, file_get_contents(__DIR__ . '/../../img/icon-text-white.svg'));
+			} elseif ($room->getObjectType() === 'share:password') {
+				$file = new InMemoryFile($token, file_get_contents(__DIR__ . '/../../img/icon-password-white.svg'));
+			} elseif ($room->getObjectType() === 'emails') {
+				$file = new InMemoryFile($token, file_get_contents(__DIR__ . '/../../img/icon-mail-white.svg'));
+			} elseif ($room->getType() === Room::TYPE_PUBLIC) {
+				$file = new InMemoryFile($token, file_get_contents(__DIR__ . '/../../img/icon-public-white.svg'));
+			} else {
+				$file = new InMemoryFile($token, file_get_contents(__DIR__ . '/../../img/icon-contacts-white.svg'));
+			}
+		}
+		return $file;
+	}
+
+	public function deleteAvatar(Room $room): void {
+		try {
+			$folder = $this->appData->getFolder('room-avatar');
+			$token = $room->getToken();
+			$folder->delete($token);
+			$this->cache->clear($token . '.avatarVersion');
+		} catch (NotFoundException $e) {
+		}
+	}
+
+	public function roomHasAvatar(Room $room): bool {
+		try {
+			$folder = $this->appData->getFolder('room-avatar');
+			if ($folder->fileExists($room->getToken())) {
+				return true;
+			}
+		} catch (NotFoundException $e) {
+		}
+		return $room->getType() === Room::TYPE_ONE_TO_ONE;
+	}
+
+	public function getAvatarVersion(Room $room, ?string $userId): string {
+		if ($room->getType() === Room::TYPE_ONE_TO_ONE && $userId) {
+			return (string) $this->config->getUserValue($userId, 'avatar', 'version', '0');
+		}
+		return (string) ($this->cache->get($room->getToken() . '.avatarVersion') ?? 0);
+	}
+
+	public function getAvatarUrl(Room $room, ?string $userId = null): string {
+		if (!$this->roomHasAvatar($room)) {
+			return '';
+		}
```

Comment:
We want to have a fallback of some sorts. If this does not always return a working image, clients need to have multiple logic paths again to decide which URL to load, etc.